### ISSUE TITLE
Layout is not respected when passed by props Continued

### ISSUE THREAD

**Issue by nathan-chappell** [STRML/react-grid-layout#1286]
I'm bringing up #870 again because I can't prevent the component from setting a new layout with the current implementation. 

## Desired behavior
If I pass a layout as props, that layout should be what appears on the screen.

## Use Case
I wish to manage the layouts on my own.  When `onLayoutChange` fires I will handle the change with my own logic (namely, some reducer in redux), and pass the updated layout to ReactGridLayout.  In some cases, the layout should not change and revert to the "current layout."

## Current behavior
Because `lodash.isEqual` is used to compare the previous layout's state and the current props layout, if I pass in the "current layout" it is ignored and a new layout is put in place as if I weren't managing it

## Relevant code:

In file [ReactGridLayout.jsx](https://github.com/STRML/react-grid-layout/blob/b4caad3af49717e2ee4d8b8dc653eef9a3a684c2/lib/ReactGridLayout.jsx#L167) these lines are preventing the desired behavior:

https://github.com/STRML/react-grid-layout/blob/b4caad3af49717e2ee4d8b8dc653eef9a3a684c2/lib/ReactGridLayout.jsx#L165-L170

## Proposed fix
Ideally, I would change `!isEqual(nextProps.layout, prevState.propsLayout)` to `nextProps.layout !== prevState.propsLayout`, and put the onus on the user to insure that shallow reference comparison will be enough to determine if the layout in props is actually a new layout or not.  However, I can see this possibly breaking some components which recreate their layout on each render (which they probably shouldn't be doing...).

I propose either the above change, or perhaps a flag prop on ReactGridLayout, such as `managed`, which would allow the user to opt-in to shallow reference comparison in the above code.  For example:

```jsx
      if (
        !isEqual(nextProps.layout, prevState.propsLayout) ||
        nextProps.compactType !== prevState.compactType ||
        (nextProps.managed && (nextProps.layout !== prevState.propsLayout))
      ) {
```

The `managed` prop should get a default value of `false` so that existing code will work as if it is not there.

If @STRML doesn't see any major issues with this, I'll make a PR implementing the change.

**Comment by STRML** [comment]
You say:
> Because lodash.isEqual is used to compare the previous layout's state and the current props layout, if I pass in the "current layout" it is ignored and a new layout is put in place as if I weren't managing it

Are you modifying the object that was passed as the original `layout`? Could you do a codesandbox (https://codesandbox.io/s/5wy3rz5z1x) that shows this issue?

**Comment by nathan-chappell** [comment]
@STRML  Thanks for your response.  I put some code together demonstrating the issue on [codebox](https://codesandbox.io/s/recursing-cannon-5sp94?file=/src/index.js).  I hope that the comments explain what is going on well enough, but the point is that, when "L1" is the selected layout, the user should not be able to change the layout.  The box can drag the boxes, but when the mouse is released it should go back to its original position.  When "L2" is selected, the user can change the layout as she pleased, and and when she goes to "L1" then back to "L2", that layout is remembered.

This behavior is precisely what occurs when `!isEqual(nextProps.layout, prevState.propsLayout)` is changed to `nextProps.layout !== prevState.propsLayout`.  I think it makes it even more clear when I consider a "case analysis matrix" comparing the use of `isEqual` to `!==`...


comparison | analysis
----|----
`isEqual` and `===` | previous props are same as new props in value and reference.  Props are likely intended to be an initial state, and ignored later
`isEqual` and `!==` | My case.  The original props and new props are same in value, but one of two things are probably true: **(a)** the user wishes this layout to be respected (my case), **(b)** the user is recreating the same props object on each render (bad practice).
`!isEqual` and `!==` | A new value and object.  Clearly this layout should be respected (and currently is).
`!isEqual` and `===` | The value has changed, but not the reference.  This means an item from the layout object was changed...  This is in violation of the principle of *immutable state* which seems to be common in react.  Note that even if you do wish to directly change state objects, you can still get `!==` by using a spread operator when passing the layout as props.

Basically, I don't understand the reasoning behind using `_.isEqual`.  It seems to disable a use case, and reference comparison seems to be more "react-y".  If there is a good reason for using `_.isEqual`, then I'm certainly curious.  Any thoughts?

**Comment by Kyusung4698** [comment]
Any update on this?

**Comment by hednowley** [comment]
@STRML I appreciate the swift response!

I would expect the grid to only obey the provided `layout`/`data-grid` props, and for `onLayoutChange` (and the other similar callback props) to only emit new “proposed” layouts. This lets the grid-consumer hold the definitive layout state and transform it however they like, eliminating the whole class of problems which arise when the layout prop differs from the grid's internal layout state. You can see this pattern illustrated in the API of Facebook’s [draft.js library](https://draftjs.org/docs/getting-started#usage).

I’m not familiar with the RGL implementation but I’ll take a peek as this particular issue breaks a use-case which it would otherwise be a great fit for. The change I mentioned would not be backwards compatible and would likely require a major version bump or a new prop flag in the meantime (e.g. `managed` as suggested above).

**Comment by philiparvidsson** [comment]
So I'm banging my head against the wall on this bug. I am doing a calendar where the user can move time segments around. Each move results in a confirmation dialog. If the user selects no, the time segment should be moved back. It breaks seemingly randomly because RGL doesn't really care about the layout I am passing into it (despite me creating a new array each time). What are some better alternatives to RGL?

**Comment by philiparvidsson** [comment]
Just as I typed my last comment, I came up with this idea which seems to work:

Just add a random property to your layout item like this:


```
{
  ...
  hack: Date.now()
}
```

It breaks the equality comparison which forces RGL to update.

**Comment by pelletier197** [comment]
Late to the discussion, but I used @philiparvidsson's solution and it works. Not super great, but it will do for now. 

I also agree with the solution by @hednowley. This is usually how you want components to work.
